File: BookUtils.java
package net.glitchtechs.starapi;

import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.chat.ClickEvent;
import net.md_5.bungee.api.chat.HoverEvent;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.api.chat.ClickEvent.Action;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.BookMeta;
import org.bukkit.inventory.meta.BookMeta.Generation;

public class BookUtils {
    public static ItemStack createBookGUI(String title, String author, int totalPages) {
        ItemStack book = new ItemStack(Material.WRITTEN_BOOK);
        BookMeta bookMeta = (BookMeta)book.getItemMeta();
        bookMeta.setTitle(title);
        bookMeta.setAuthor(author);
        bookMeta.setGeneration(Generation.ORIGINAL);

        for(int i = 1; i <= totalPages; ++i) {
            bookMeta.addPage(new String[]{String.valueOf(i)});
        }

        book.setItemMeta(bookMeta);
        return book;
    }

    public static void setPageContent(BookMeta bookMeta, int pageNumber, String clickUrl, String hoverText, String page, String... lines) {
        if (pageNumber >= 1 && pageNumber <= 50) {
            BaseComponent[] components = TextComponent.fromLegacyText(MessageUtils.hex(String.join("\n", lines)));
            ClickEvent clickEvent;
            if (clickUrl != null && !clickUrl.isEmpty()) {
                clickEvent = new ClickEvent(Action.OPEN_URL, clickUrl);
                components[components.length - 1].setClickEvent(clickEvent);
            }

            if (hoverText != null && !hoverText.isEmpty()) {
                HoverEvent hoverEvent = new HoverEvent(net.md_5.bungee.api.chat.HoverEvent.Action.SHOW_TEXT, TextComponent.fromLegacyText(MessageUtils.hex(hoverText)));
                components[components.length - 1].setHoverEvent(hoverEvent);
            }

            if (page != null && !page.isEmpty()) {
                clickEvent = new ClickEvent(Action.CHANGE_PAGE, page);
                components[components.length - 1].setClickEvent(clickEvent);
            }

            bookMeta.spigot().setPage(pageNumber, components);
        } else {
            throw new IllegalArgumentException("Invalid page number " + pageNumber);
        }
    }
}

File: Kotlin/src/BookUtils.kt
package net.glitchtechs.starapi

import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.BookMeta


object BookUtils {
    fun createBookGUI(title: String?, author: String?, totalPages: Int): ItemStack {
        val book = ItemStack(Material.WRITTEN_BOOK)
        val bookMeta = book.itemMeta as BookMeta?
        bookMeta!!.setTitle(title)
        bookMeta.author = author
        bookMeta.generation = BookMeta.Generation.ORIGINAL

        for (i in 1..totalPages) {
            bookMeta.addPage(*arrayOf(i.toString()))
        }

        book.setItemMeta(bookMeta)
        return book
    }

    fun setPageContent(bookMeta: BookMeta, pageNumber: Int, clickUrl: String?, hoverText: String?, page: String?, vararg lines: String?) {
        if (pageNumber >= 1 && pageNumber <= 50) {
            val components = TextComponent.fromLegacyText(MessageUtils.hex(java.lang.String.join("\n", *lines)))
            var clickEvent: ClickEvent
            if (clickUrl != null && !clickUrl.isEmpty()) {
                clickEvent = ClickEvent(ClickEvent.Action.OPEN_URL, clickUrl)
                components[components.size - 1].clickEvent = clickEvent
            }

            if (hoverText != null && !hoverText.isEmpty()) {
                val hoverEvent = HoverEvent(HoverEvent.Action.SHOW_TEXT, TextComponent.fromLegacyText(MessageUtils.hex(hoverText)))
                components[components.size - 1].hoverEvent = hoverEvent
            }

            if (page != null && !page.isEmpty()) {
                clickEvent = ClickEvent(ClickEvent.Action.CHANGE_PAGE, page)
                components[components.size - 1].clickEvent = clickEvent
            }

            bookMeta.spigot().setPage(pageNumber, *components)
        } else {
            throw IllegalArgumentException("Invalid page number $pageNumber")
        }
    }
}